Freebsd unit test fixes

bazsi opened the pull request:
No NEWS file is needed, especially until the fixes in this branch are confirmed.

@yurivict I couldn't reproduce the test_python_logmsg breakage on my local FreeBSD box (e.g. as you reported it in #4230), but while testing that I've found an unrelated issue that I've fixed in this branch. 

Can you please check if it fixes both issues in your environment? Thanks.

bazsi commented:
@czanik can you please check these unit test problems on FreeBSD? I've tried to reproduce the issue but my FreeBSD skills are somewhat limited. Thanks

czanik commented:
@bazsi ```make check``` is not part of the FreeBSD ports Makefile, and I never compiled syslog-ng without ports. Now updating my build machine and trying to figure out what @yurivict did. It seems that he did a mixture of using ports and compiling by hand.

czanik commented:
@bazsi @yurivict  OK. Here is what did. In sysutils/syslog-ng first I did a ```make check-orphans```, the command I use to follow up with packaging list changes. Once it finished I changed to work/syslog-ng-3.38.1 and did a ```gmake check```. It failed, saying that unit tests are disabled. It turned out that ```criterion``` is now packaged for FreeBSD, so I installed it and repeated the process:
```
FAIL: modules/timestamp/tests/test_date
```
So, with 3.38.1 release a single test failed. As far as I can see, your current patch is related to time.

It seems that if I want to integrate ```make check``` support to the port, I'll have to edit the package list as well. But first I need to figure out if it can be made optional.

Before changing anything in the port, I check your fix on an up-to-date git snapshot.

bazsi commented:
Yeah, my change in this pr should fix test_date

czanik commented:
OK, test_date is fixed. However, there is now another fail:
```
FAIL: modules/python/tests/test_python_logmsg
```
Could it be related to --disable-python-modules, which is necessary not to look for Kubernetes Python modules?

bazsi commented:
Can you run this separately, outside of make check? And paste the output
here.

Also, it might be needed to log into that environment if possible.


On Mon, Dec 12, 2022, 17:20 Peter Czanik ***@***.***> wrote:

> OK, test_date is fixed. However, there is now another fail:
>
> FAIL: modules/python/tests/test_python_logmsg
>
> Could it be related to --disable-python-modules, which is necessary not to
> look for Kubernetes Python modules?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/4232#issuecomment-1346823136>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5XCKSNR2K2S34KKMTLWM5GGVANCNFSM6AAAAAASUAOJK4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

czanik commented:
```
root@fb131:~/syslog-ng/work/syslog-ng-3.38.1.358.g3093f4e # ./modules/python/tests/test_python_logmsg
[----] modules/python/tests/test_python_logmsg.c:210: Assertion Failed
[----]   
[----]   The expression (as strings) (value) == (params->expected_log_msg_value) is false: actual=`1664887692.123` expected=`1664894892.123`.
[----]   
[FAIL] python_log_message::test_python_logmessage_set_value: (0.00s)
[====] Synthesis: Tested: 18 | Passing: 17 | Failing: 1 | Crashing: 0 
```

czanik commented:
Oops, wrong button... Reopening.

bazsi commented:
Thanks. This seems to be time related too. There are roughly 2 hours
between those timestamps, which is the amount between CET and GMT.

On Mon, Dec 12, 2022, 18:39 Peter Czanik ***@***.***> wrote:

> ***@***.***:~/syslog-ng/work/syslog-ng-3.38.1.358.g3093f4e # ./modules/python/tests/test_python_logmsg
> [----] modules/python/tests/test_python_logmsg.c:210: Assertion Failed
> [----]
> [----]   The expression (as strings) (value) == (params->expected_log_msg_value) is false: actual=`1664887692.123` expected=`1664894892.123`.
> [----]
> [FAIL] python_log_message::test_python_logmessage_set_value: (0.00s)
> [====] Synthesis: Tested: 18 | Passing: 17 | Failing: 1 | Crashing: 0
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/4232#issuecomment-1346938053>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5V4BOXOXR7DOUQCAVLWM5PLJANCNFSM6AAAAAASUAOJK4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bazsi commented:
This should fix this:

```
diff --git a/modules/python/tests/test_python_logmsg.c b/modules/python/tests/test_python_logmsg.c
index 4f632faa3..dfc4b96a2 100644
--- a/modules/python/tests/test_python_logmsg.c
+++ b/modules/python/tests/test_python_logmsg.c
@@ -95,6 +95,7 @@ _construct_py_parse_options(void)
 void
 setup(void)
 {
+  setenv("TZ", "UTC", TRUE);
   app_startup();
 
   init_parse_options_and_load_syslogformat(&parse_options);

```

bazsi commented:
The crashes sometimes are expected, when the tests exercise paths that
should abort, but I'll check those.

For me a threaded testcase hung when I last tried it. the one that tests
wakeups. Isn't a test program running in the background when this is hung?

On Thu, Dec 15, 2022, 09:28 Peter Czanik ***@***.***> wrote:

> I updated to the latest git snapshot and also updated FreeBSD ports and
> packages to the very latest version. Unfortunately gmake check now hangs
> here:
>
> PASS: lib/transport/tests/test_transport_factory_registry
> PASS: lib/transport/tests/test_multitransport
>
> In dmesg I see a couple of crashes:
>
> pid 2111 (test_consecutive_ac), jid 0, uid 0: exited on signal 6 (core dumped)
> pid 2118 (test_consecutive_ac), jid 0, uid 0: exited on signal 6 (core dumped)
> pid 2259 (test_transport_fact), jid 0, uid 0: exited on signal 6 (core dumped)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/4232#issuecomment-1352714587>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5S2SVKMBSRAOTMWODDWNLJCFANCNFSM6AAAAAASUAOJK4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

czanik commented:
```
root   2056   0.0  1.8 31460 18296  0  I+   11:49    0:00.65 gmake --no-print-directory check-TESTS
root   2063   0.0  0.3 13596  2960  0  I+   11:50    0:00.00 /bin/sh -c set +e; bases='libtest/tests/test_mock_transport.log lib/ack-tracker/tests/test_consecutive_ack_record_container.log lib/ack-tracker/tests/test_instant_ack_tracker.l
root   2069   0.0  1.7 26340 17316  0  I+   11:50    0:00.60 gmake --no-print-directory test-suite.log TEST_LOGS=libtest/tests/test_mock_transport.log lib/ack-tracker/tests/test_consecutive_ack_record_container.log lib/ack-tracker/tests/
root   2262   0.0  0.3 13596  2892  0  I+   11:52    0:00.00 /bin/sh ./test-driver ./lib/logproto/tests/test_logproto
root   2269   0.0  1.2 33232 12520  0  S    11:52    0:00.11 ./lib/logproto/tests/test_logproto
root  61855   0.0  1.4 26340 14152  0  I+   11:48    0:00.43 gmake check
root  61857   0.0  1.4 26340 14004  0  I+   11:48    0:00.45 gmake --no-print-directory check-recursive
root  61859   0.0  0.3 13596  2864  0  I+   11:48    0:00.00 /bin/sh -c fail=; \\\nif (target_option=k; case ${target_option-} in ?) ;; *) echo "am__make_running_with_option: internal error: invalid" "target option '${target_option-}' sp
root  61864   0.0  1.7 26340 17108  0  I+   11:48    0:00.52 gmake --no-print-directory check-am

```

czanik commented:
Well. Interesting. I had a couple of meetings, and I left ```gmake check``` running. It turned out that it not hung, only it's extremely slow. Normally running all tests finish in my vm in less than a minute. Now about 50 tests took more than an hour. Neither the host, nor the vm has a high load.

czanik commented:
It took a couple of hours, but both failing tests are green now.

MrAnno commented:
@czanik Is it always `./lib/logproto/tests/test_logproto` the one that goes into an interruptible sleep for a long time?

czanik commented:
@MrAnno  That's the first one. But after that many others as well. But just a week earlier everything was fast. So, I would not be surprised if everything worked again quickly after the next port upgrade.

czanik commented:
@bazsi @MrAnno Did a ports update on my FreeBSD build machine. Python and criterion dependencies got updated. Tests suddenly finish in less than a minute in my VM. With this patch all tests are green. :+1:

czanik reviewed (approved):
All tests are now green on FreeBSD

MrAnno commented:
I think we can merge this for v4.0.
I'm changing the milestone. Let me know if I overlooked something.
